### DISCUSSION TITLE
Your dev-environment?

### DISCUSSION THREAD

**Discussion by sreichel** [OpenMage/magento-lts#2466] [Polls]
Good evening :)

After having some difficulties at #2400, i would be interested to know what your development environments look like.

Win/MacOS/Linux?
Composer Y/N?
Local db/webserver/..., docker or ddev?

**Comment by fballiano** [comment]
MacOS with apache/redis/mysql/modphp (multiple versions) installed via homebrew

EDIT: composer, phpstorm

**Comment by kiatng** [comment]
Mine is probably very strange. IDE; Windows, VS Code, auto-FTP to Linux dev server for testing, git deployment script to staging/production. No composer.

**Comment by addison74** [comment]
In the development environment I use Debian 10 as operating system. I have several virtual domains running PHP 7.3 and 8.1. I flirted for a while with Ubuntu, which is derived from Debian, and with CentOS. At the beginning in 2010 I tested Magento in Windows with WAMP and XAMPP, I remember that it was very slow, which is why I was with the X-Cart team for 10 years and then CS-Cart for 4 years to return to Magento in 2014, when it was no longer a problem to get a dedicated server with good resources.

I don't use Docker, Composer, but I can do it anytime if necessary. As IDE I use Phpstorm. As an editor, VS Code - especially on the CSS/SASS, JS, Node, which has many facilities and I feel better using it. I use Git in the terminal and GitHub Desktop for simplicity in the relationship with GitHub, especially to test PRs.

**Comment by Sdfendor** [comment]
Windows WSL (ubuntu 18.04): nginx, PHP 8.1-fpm (8.0 available but moving away from it, 8.2 incoming), Maria DB, Solr 8 besides Solr which runs in windows all is running within WSL.

**Comment by sreichel** [comment]
Interessting so far ... 4 comments, 4 different setups :)

I started with Win + Xampp + (Notepad++, Putty, Firefox-FTP-Plugin, tortoiseGit/later Sourcetree). Performance was bad, it was unflexible and i missed a few linux commands. Using VirtualBox with Ubuntu 16.04 (?) was an update, but it was still hard to run different php-versions for different projetect. I had to switch for every project. Docker seemed good, but i've never really worked with it ... 

I'm still on Win7 + VBox + Ubuntu, but best decission i made, was to spent half a hour, to get familiar with DDEV.

- it supports Magento1/OM
- there are containers for everything (Varnish, Solr, Redis, crontab, ... MySql/MariaDB/..., apache/nginx)
- you can run projects with different php/webserver/db at the same time (!!!)
- prebuilt/preinstalled tools like npm and phpstorm support (setting up SCSS filewatch in phpstorm takes a minute)
- support for Xdebug, Xhprof and blackfire.io
- ...

Tools? Only phpStorm - and tmux :)

**Reply by addison74** [reply]
DDEV is really a good option. In the last two days I have seen tutorials on YouTube of its creator and I can say that it is worth considering. In Windows it is a better option than XAMPP/WAMP for speed. The recommendation is based on WSL2, otherwise the difference is not too big. It requires a little patience during installation, following a list of steps from the official website, but it is a good way to avoid using Hyper-V/VirtualBox/VMWare. Those who already use Docker can take another step towards DDEV.

**Comment by justinbeaty** [comment]
I am just running a simple PHP + MySQL + nginx environment on Ubuntu 20.04. Just emacs as a text editor so I don't get a lot of the fancy IDE support unless I spend time configuring it.

Composer: yes, although for deployment I build into a git repo including vendor and just deploy that. Much better than relying on composer in production IMO.

**Reply by sreichel** [reply]
> Just emacs as a text editor so I don't get a lot of the fancy IDE support

Does emacs support php auto-completion? Just found https://github.com/emacs-php/auto-complete.

Do you have any benefit of added docblocks/annotations?

**Reply by justinbeaty** [reply]
I have basic autocompletion for PHP built-ins, but nothing that integrates with the docblocks. I've been considering trying PHPStorm to take advantage of all the improvements being merged into this repo. I saw you posted how to get a free license for F/OSS development, although I'm not positive I qualify due to the "Licenses are provided only to project leads and core project committers" clause. Either way it's just hard to break 10+ years of muscle memory from using emacs. I know many IDEs have emacs-like keybindings, but it's never the same.

**Comment by colinmollenhour** [comment]
Personal OS: Windows 10/11
IDE: PhpStorm
Composer: Yes (via Docker only)
Webserver: Linux via WSL2 + Docker for Windows + Apache and nginx+php-fpm

I put all project files in the WSL2 filesystem so, to the docker containers, they appear to be on a native Linux filesystem. PhpStorm runs in Windows and accesses them via the `\\wsl.localhost\` share for Windows 11 or `\\wsl$` on Windows 10.

I have refused to install PHP on any system in the last 4+ years since switching to Docker and never looked back. It is very well supported by all operating systems and all major tools like PhpStorm and VSCode. I can use any PHP version for any project just by swapping base container version numbers this way. I believe using Docker to run web servers both on your local dev environment and production is vastly superior to the old ways of installing it directly on the host whether you are on Windows, MacOS or Linux.

Docker is basically a must-know tool right up there with git so I'd love to see everyone accept Docker as the de-facto way to start a web server on any platform. Wrapping it with tools like `ddev` can be very handy but I would say that is up to personal preference. I find using `docker-compose` to be perfectly sufficient in all cases.

**Comment by elidrissidev** [comment]
I'm a full time penguin:

OS: Manjaro KDE (Arch-based)
Web server: apache
PHP: php-fpm 8.1 and 7.4
Database: MariaDB, along with phpMyAdmin
Editor: Been using mostly VSCode, ~but might give PhpStorm et al a try if I qualify for an OSS license~ now using PhpStorm.
Other software: Composer, mailhog.

**Reply by sreichel** [reply]
Did you qualify for OSS license? Could be interesting for others :)

**Reply by elidrissidev** [reply]
Yes I did!

**Comment by Flyingmana** [comment]
Mine is primarily Windows/WSL2

did use a while ddev, but had some conflicts with its router and my kubernetes environment.
Most of the time I use docker-compose then.

For some smaller things I also use the web gitpod environment, wanted to try the new local one of it next.

**Comment by alexh-swdev** [comment]
Windows w/ PHPStorm and "sync on save" to a ssh-network-share to the raspbian-testserver.

**Comment by addison74** [comment]
Since my last post at the end of August, I had an interesting experience, in the sense that I was asked for a basic test environment in Windows (with which I don't really interact) based on XAMPP and I ended up discovering DDEV. 

Initially I used the combination of Docker + DDEV + Phpstorm all in Windows to then use (Docker in Windows) + (DDEV + WSL2 + Debian from Store) + (Phpstorm in Windows). 

The first version even with mutagen enabled runs very slow, starting DDEV sometimes took 3-4 minutes and the loading time of the pages what similar to XAMPP. The second version is close in terms of performance to what I have been running for years in (Ubuntu + Virtualmin), that is, everything is extremely fast. 

I recommend everyone to try DDEV because there is a huge improvement for creating a testing environment, especially in Windows. I am not an avid user of Docker, but I have spent a lot of time with it. What Randy Fay did in DDEV was to stop getting our hands dirty in configurations, and everything should be very fast even for a novice. Mkcert comes to provide an accepted certificate for HTTPS connections, we can switch easy from apache to nginx, from mariadb to percona, from php 7.0 to 8.2, phpMyAdmin, Mailhog and many more. 

Maybe we should make a thread where we can share our experiences related to DDEV, anyone can get a testing environment for OpenMage in an extremely short time.

**Reply by addison74** [reply]
Before starting with DDEV you need to evaluate your hardware. Here are some figures related to memory usage in **Windows 10 Pro** latest version.

**Memory in use = 2.2 GB (initial)**

+ Running Docker Desktop = 4.5 GB (Vmmem process takes 1.75 GB to run, but it can be limited to a specified value by editing .wslconfig file)
+ Running Ubuntu = 4.6 GB
+ Running ```ddev start``` command inside Ubuntu = 5.4 GB
+ Running ```ddev launch``` (opens the Frontend in Chrome) = 5.9 GB
+ Opening Backend in browser + 3 more tabs = 6.3 GB
+ Running Phpstorm = 8.7 GB

In conclusion, you need 16 GB in your systems to stay comfortable and run properly a test environment based on the above configuration. Phpstorm can be replaced by a popular text editor like Sublime Text to run on a system with 8 GB of RAM.

What is interesting is the fact that in 2014 I worked on an online store on Magento 1.9 in a system with 8 GB of RAM. I used Windows 7 + VMware and in the VM I used Debian 7 + Virtualmin to simulate a vhost in production. I had no issues with lack of resources, I think I still had 2 GB free. Now I run in a system with 128 GB of RAM with NVeM and I look back with admiration at the resource limitations of the past. I am part of the generation that started programming on an 80286 long time ago.

**Reply by sreichel** [reply]
> In conclusion, you need 16 GB in your systems to stay comfortable

Never looked for this, but yes ... i've 16GB RAM and 10GB available for VBox. With browser opened in Win7 with 100s of tabs, some other tools, vbox + phpstorm + browser takes 15GB oO

I should think about an upgrade :)

**Reply by colinmollenhour** [reply]
ddev is mentioned in the readme [here](https://github.com/OpenMage/magento-lts/#development-environment-with-ddev), although it is fundamentally just a wrapper around Docker. Have you tried the single-command install `dev/openmage/install.sh`?

**Reply by addison74** [reply]
DDEV is a real help when configuring Docker can be intimidating. You may take a look and you will find that they have studied enough to create configurations that cover a lot of CMS's. I haven't tried that command, but I promise I will. I would like to do an installation Docker + DDEV + Phpstorm in a popular Linux distribution to check what memory resources it uses.

**Reply by colinmollenhour** [reply]
The install.sh route is much lighter/simpler than ddev. If you already have Docker and bash installed (e.g. WSL2) then you can literally clone the repo and have a working environment in one command and in about 30 seconds:

![image](https://user-images.githubusercontent.com/38738/193682988-f5e3b255-fa64-4639-9ef8-dde7ae1fdaa1.png)

**Reply by addison74** [reply]
Here are some figures related to the memory usage in **Ubuntu 20.04.1 LTS**.

**Memory in use = 1.6 GB (initial)**
(Docker runs as service and takes only 39 MB from 1.6 GB)

+ Running ```ddev start``` command inside Ubuntu = 2.0 GB
+ Running ```ddev launch``` command (opens the Frontend in Chrome) = 2.3 GB
+ Opening Backend in browser + 3 more tabs = 2.6 GB
+ Running Phpstorm = 4.7 GB

Comparing with Windows 10 Pro, the memory usage in Ubuntu is half as size. I was impressed by the speed (loading pages in browser, Phpstorm indexing). Docker really makes a great job.

**Reply by colinmollenhour** [reply]
@ADDISON74 @sreichel @elidrissidev @Flyingmana @fballiano and others..

Regarding the issues with `install.sh` (#2647) do you think `install.sh` is worth keeping or do you think it should be abandoned in favor of making ddev the only supported dev environment? Personally I prefer the simple/lightweight environment provided by a simple docker-compose project but if I'm the only one then there is no reason to keep it in the core. 😄

**Reply by elidrissidev** [reply]
@colinmollenhour Although I'm not currently using it, I'd prefer using Docker over DDEV just for simplicity, and your script is a great time-saver. So +1 from me for keeping it.

**Reply by addison74** [reply]
@colinmollenhour - I love Docker but I don't like to create configurations from scratch. If it wasn't for @sreichel to make me curious about DDEV, I wouldn't have tried it a month and a half ago and I wouldn't even have come across your script these days. I went through most of the situations that an average-advanced user would face, I also found a few issues, I appreciated the performances Docker offers in WIndows+WSL2 and Ubuntu. You asked a natural question and in order to give you an honest answer I should put down on paper what I found pros and cons in each solution. I like the simplicity but at the same time I appreciate the features offered.

My opinion is the following, DDEV is the way to go to provide a fast test and development environment for OpenMage. The advantages are many. I would mention just a few, it has many customization facilities, recognizes SSL certificates and can create LE certificates, useful commands (https://ddev.readthedocs.io/en/latest/users/basics/cli-usage/), the ease of change the webserver, database, PHP, quickly in a file, snapshots, export/import database, can be added quickly through Varnish, Redis and many others needed. 

Regarding your script, it is a pity to remove it because that is where your experience with Docker can be found, please update it regarding the images it uses, check the configuration and maintain it for the future if you can. It does not have as many facilities as DDEV, but it remains a research topic for anyone who is interested.

**Reply by colinmollenhour** [reply]
I definitely agree ddev is much more robust and for devs that want to take the time to learn its many features it is a good solution and I am not advocating in any way to remove it. I'm just wondering if that leaves a place for the more basic approach in the core or not.

There is really just one advantage of the `install.sh` script: if you just have docker and bash (which is most people now) and want to get the code running as quickly and simply as possible, it is the quickest way. For me this is enough, but I was just curious if I was the only one using `dev/openmage`. It has been slightly broken and dated (PHP 7.3) so that makes me think it doesn't get much use, although it is also not advertised much. I'm trying to find a simple container that is well maintained to replace the meanbee container.. maybe even using the ddev container would work. 😄

**Reply by addison74** [reply]
As it appears from this discussion only @sreichel and @Flyingmana uses/used DDEV, the rest of us do not even use Docker. It was natural for that script of yours to go unnoticed. Until now I have worked with VM snapshots on desktops and I have not missed anything, but seeing that some of you have promoted Docker I said to test this option as well, which I found extremely useful and I did not hesitate to express my satisfaction for DDEV. The author of the project really did a great job of providing an umbrella for the most popular frameworks/CMS's, as he included us this is very good for OpenMage.

I would call on your experience with Docker to solve an issue I identified in DDEV. I also asked for a vote here https://github.com/OpenMage/magento-lts/discussions/2648 too.

**Reply by sreichel** [reply]
@colinmollenhour i dont use it. DDEV is more convenient for my needs. I would keep it for all that use plain docker-compose. (can we rename directory from `openmage` to `docker`? :sunglasses: )

**Reply by colinmollenhour** [reply]
@sreichel Docker Compose uses the parent directory by default as the "project name" so naming the directory "openmage" was at the time the simplest/most reliable way to make that the default project name. That is, the containers will be named "openmage_mysql_1" instead of "docker_mysql_1". However it seems now there is support for a ["name" top-level element](https://docs.docker.com/compose/compose-file/#name-top-level-element) to set the project name so it could be renamed if that was added.

**Comment by addison74** [comment]
For those who want to test DDEV in Windows, I recommend installing Docker and DDEV inside Ubuntu or another used Linux distribution. Only mkcert can be installed in Windows to create and add the SSL certificate that allows the use of HTTPS connections without warning in the browser. Installing DDEV in Windows even with Mutagen enabled is like the speed far below XAMPP and far from a Linux installation. After seeing what (Docker + DDEV + Phpstorm) can do in Ubuntu installed in a VMware, I will never make a development environment in (Windows + WSL2) unless I am forced by circumstances.

**Comment by addison74** [comment]
After a year and a half of using **[Windows 10 + mkcert + WSL2 + Docker + DDEV + PHPStorm]** I can say that I have found the perfect combination for testing OpenMage with which I feel the best. It can also be implemented in a Linux distribution by removing the part related to (WSL + Linux), but I still need useful programs in Windows. I recently published my experience with DDEV here

- https://github.com/OpenMage/magento-lts/discussions/3839
- https://github.com/OpenMage/magento-lts/blob/main/docs/DDEV.md

As time changes habits, **please vote again** and don't forget to recommend the DDEV test environment to everyone, because that's the only way they will be convinced of the OpenMage advantages, which is a powerful and competitive e-commerce platform for nowadays💖💖💖.

**Reply by colinmollenhour** [reply]
The ddev setup is certainly very robust and the `docs/DDEV.md` documentation is good, so it should probably be linked in the README here: https://github.com/OpenMage/magento-lts?tab=readme-ov-file#development-environment-with-ddev

I don't think we need to settle/vote/agree on just one environment being the only way to do development, though.
- Some people just don't care about ddev and want to use their own methods (like using a remote server to run the code)
- The ddev setup is a bit heavy for first-timers and can conflict with other tools on your system - for me personally it's a bit overkill - I did try it so not speaking without experience
- The ddev environment is not at all similar to a production environment and if you have an advanced production environment you probably have your own stack to go with it and at that point ddev is more of a hassle

I think the current state is pretty good, what do you suggest changing about it?
- We recommend ddev in the README and have a detailed document for it - full featured dev environment "out of the box" essentially
- We have a one-step dev setup in dev/openmage - a good gateway to getting started - can upgrade to ddev later if it's not enough
- The one-step environment in dev/openmage is also a good blueprint for a sound production environment as of #1209
- EDIT: We also have the gitpod setup - sorry, I have not tried it yet

**Reply by addison74** [reply]
Please note the unchanged survey is not about DDEV. After 2 years since an opinion was requested by the author, there are new proposals for a test environment. Maybe those who voted in the past have changed their options. It is my case, from a Linux production environment I ended up testing OpenMage in Windows. Lately I have become an advocate for DDEV for the simple reason that OpenMage can be tested by those interested with minimal effort using the tutorial I published. I did not intend to make this discussion a pro/con for DDEV. Apart from me and Sven, I don't thing there is anyone else who uses DDEV.

**Reply by Sdfendor** [reply]
> Maybe those who voted in the past have changed their options

Switched from the setup described in https://github.com/OpenMage/magento-lts/discussions/2466#discussioncomment-3455086 to a (custom) DDEV setup living within WSL 2. Custom in the sense that it isn't the one provided by OpenMage. Furthermore we are in the process of adopting PHP 8.3.

**Comment by sreichel** [comment]
> Please note the unchanged survey is not about DDEV. After 2 years since an opinion was requested by the author, there are new proposals for a test environment. Maybe those who voted in the past have changed their options

Close this as outdated.
